libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.cpp

```diff
@@ -81,16 +93,145 @@ void AP_BattMonitor_SMBus_Maxell::timer()
 {
-    uint8_t buff[2];    // buffer to hold results
+    // buffer to hold results (1 extra byte returned holding PEC)
+    uint8_t read_size = 2 + (_pec_support ? 1 : 0);
+    uint8_t buff[read_size];    // buffer to hold results
 
     // read three bytes and place in last three bytes of buffer
     if (!_dev->read_registers(reg, buff, sizeof(buff))) {
         return false;
     }
 
+    // check PEC
+    if (_pec_support) {
+        uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, 2);
+        if (pec != buff[2]) {
+            return false;
+        }
+    }
+
     // convert buffer to word
     data = (uint16_t)buff[1]<<8 | (uint16_t)buff[0];
 
     // return success
     return true;
 }
+
+// read_block - returns number of characters read if successful, zero if unsuccessful
+uint8_t AP_BattMonitor_SMBus_Maxell::read_block(uint8_t reg, uint8_t* data, bool append_zero) const
+{
+    // get length
+    uint8_t bufflen;
+    // read byte (first byte indicates the number of bytes in the block)
+    if (!_dev->read_registers(reg, &bufflen, 1)) {
+        return 0;
+    }
+
+    // sanity check length returned by smbus
+    if (bufflen == 0 || bufflen > READ_BLOCK_MAXIMUM_TRANSFER) {
+        return 0;
+    }
+
+    // buffer to hold results (2 extra byte returned holding length and PEC)
+    uint8_t read_size = bufflen + 1 + (_pec_support ? 1 : 0);
+    uint8_t buff[read_size];
+
+    // read bytes
+    if (!_dev->read_registers(reg, buff, read_size)) {
+        return 0;
+    }
+
+    // check PEC
+    if (_pec_support) {
+        uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, bufflen+1);
+        if (pec != buff[bufflen+1]) {
+            return 0;
+        }
+    }
+
+    // copy data (excluding PEC)
+    memcpy(data, &buff[1], bufflen);
+
+    // optionally add zero to end
+    if (append_zero) {
+        data[bufflen] = '\0';
+    }
+
+    // return success
+    return bufflen;
+}
+
+// get PEC support with the version value in SpecificationInfo
+bool AP_BattMonitor_SMBus_Maxell::get_pec_support()
+{
+    uint16_t data;
+    uint8_t buff[READ_BLOCK_MAXIMUM_TRANSFER + 1];
+
+    // specification info
+    if (read_word(BATTMONITOR_SMBUS_MAXELL_SPECIFICATION_INFO, data)) {
+        if ((data & 0xFF) == BATTMONITOR_SMBUS_11_PEC_SUPPORT) {
+            if (read_block(BATTMONITOR_SMBUS_MAXELL_MANUFACTURE_NAME, buff, true)) {
+                // In Hitachi maxell battery, specification info is 0x31 (SBSv1.1 with PEC support) but PEC isn't support
+                if (strcmp((char*)buff, "Hitachi maxell") == 0) {
+                    _pec_confirmed = true;
+                    _pec_support = false;
+                    return 1;
+                }
+            }
+            _pec_support = true;
+            // Check if it can get the correct value with PEC support
+            if (read_word(BATTMONITOR_SMBUS_MAXELL_SPECIFICATION_INFO, data)) {
```

Comment:
It is the same condition as Line 171. Since the value of data has not changed, I think that this judgment is unnecessary.